### PR TITLE
kvserver: add logStore type with storeEntries method

### PR DESCRIPTION
**Pull request by pav-kv**
This change factors out a `logStore` type stub as a foundation for a separated log storage, and moves the log entries persistence code to be its `storeEntries` method.

Epic: CRDB-220
Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/client_raft_test.go
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/client_raft_test.go:608]: Oups, a false positive from GoLand auto-refactoring.

**Review by tbg — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1, 7 of 7 files at r2, 1 of 1 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91742)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

___
*[`pkg/kv/kvserver/raft_log_queue.go` line 267 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGbJV2UB-X87mAsX_UB:-NGbJV2UB-X87mAsX_UC:b-9je3a4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8624be84ee6cb2440b8836893a36099ea751d83f/pkg/kv/kvserver/raft_log_queue.go#L267)):*
> ```Go
> 	pendingSnapshotIndex := r.getSnapshotLogTruncationConstraintsRLocked(anyRecipientStore)
> 	lastIndex := r.mu.lastIndex
> 	// NB: byteSize above adjusts for pending truncations that have already
> ```

I think you realized this in another comment already but just in case, this seems to be a stray rename. Though maybe instead of changing the comment back you want to rename the local `raftLogSize` here, your call.
___
*[`pkg/kv/kvserver/replica_raft.go` line 1082 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGbK0YOE_O-Fu5m9a-f:-NGbK0YP8kueR845SsDD:bus55c6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bab8518c963da95cfe716b803aea07471e16289/pkg/kv/kvserver/replica_raft.go#L1082)):*
> ```Go
> }
> 
> // storeCommittedEntries persists new raft-committed Entries to the log storage.
> ```

These entries are not necessarily committed. Here, we are appending to the log; the entries may or may not be committed or they may be committed and we don't know it. We really don't care; we want to make them durable.

I'm surprised you're handing the `Ready` to this method. I expected to see only the `Entries` and `HardState` fields. We should also not construct the batch and also not have this anchored on `*Replica`. I see that you're moving in that direction in the next commit so as an intermediate state I'm fine with this as is, just making sure we're going to follow through on this swiftly.

I think it would make sense to polish up the first three commits, augment the third a bit to get the standalone method, and then do a separate PR that does the "new package" stuff, WDYT?
___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 683 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGbJ33N8BZVgHx9qjBB:-NGbJ33N8BZVgHx9qjBC:b-2x3daz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1c219f2e5c161252787e0c560aca8cb67ff90a3/pkg/kv/kvserver/replica_raftstorage.go#L683)):*
> ```Go
> 		}
> 	}
> 	prev.raftLogSize = prev.raftLogSize + diff.SysBytes
> ```

There seems to be a bug here, you ought to be setting `newState.raftLogSize`. I'm always wary of populating fields piecemeal, so I would `return raftLogState{...}` at the end of this methods and use locals where necessary, i.e. `newLastIndex` (you don't need `newLastTerm`).

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91742)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/replica_raft.go` line 1082 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGbK0YOE_O-Fu5m9a-f:-NGbalwo9UM6qEVIpfCe:bvsli1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bab8518c963da95cfe716b803aea07471e16289/pkg/kv/kvserver/replica_raft.go#L1082)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

These entries are not necessarily committed. Here, we are appending to the log; the entries may or may not be committed or they may be committed and we don't know it. We really don't care; we want to make them durable.

I'm surprised you're handing the `Ready` to this method. I expected to see only the `Entries` and `HardState` fields. We should also not construct the batch and also not have this anchored on `*Replica`. I see that you're moving in that direction in the next commit so as an intermediate state I'm fine with this as is, just making sure we're going to follow through on this swiftly.

I think it would make sense to polish up the first three commits, augment the third a bit to get the standalone method, and then do a separate PR that does the "new package" stuff, WDYT?
</blockquote></details>

Good catch, renamed to `storeLogEntries`.

Yes, it's an intermediate stage. Handing in `Ready` is to keep this change mechanical at first, the TODO in the method comment captures the intention to move away from `Ready` and pass in things only needed for the log.

Why not construct the batch? This func constructs, populates, commits, and closes the batch (so looks self-contained). Not sure what will be left of it if I move the batch away. Are you alluding to having something like `DoRaftMutation` [here](https://github.com/sumeerbhola/cockroach/blob/787e4103d54d1c5ee88249bacbb4a85c9d44801b/pkg/kv/kvserver/replicas_storage.go#L728)? For now I don't have a good feeling how this interface reflects what we're doing in this func. If we push most of the stuff above the interface anyway then I'm not sure what the interface implementation would need to do - just commit the batch?

I think it's nice to have a self-contained function that knows how to construct a batch - someone needs to do it anyway. If it's necessary to integrate with the interface, we can do it later, I think it's a bit early and unclear now.

Polishing done, PTAL.
___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 683 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGbJ33N8BZVgHx9qjBB:-NGbabEC8vKes9KV7eV6:b8gj3mz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1c219f2e5c161252787e0c560aca8cb67ff90a3/pkg/kv/kvserver/replica_raftstorage.go#L683)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

There seems to be a bug here, you ought to be setting `newState.raftLogSize`. I'm always wary of populating fields piecemeal, so I would `return raftLogState{...}` at the end of this methods and use locals where necessary, i.e. `newLastIndex` (you don't need `newLastTerm`).
</blockquote></details>

Good catch, fixed.
___
*[`pkg/kv/kvserver/raft_log_queue.go` line 267 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGbJV2UB-X87mAsX_UB:-NGba-Y_B7P9TpYrcjLZ:b-2ft4zs) ([raw file](https://github.com/cockroachdb/cockroach/blob/8624be84ee6cb2440b8836893a36099ea751d83f/pkg/kv/kvserver/raft_log_queue.go#L267)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I think you realized this in another comment already but just in case, this seems to be a stray rename. Though maybe instead of changing the comment back you want to rename the local `raftLogSize` here, your call.
</blockquote></details>

Reverted accidental comment changes. Local variables (out of the struct context) make more sense this way. The struct field has a luxury of being shortened because it's in the context of a "raft log" struct anyway.

**Comment by pavelkalinnikov** [comment]
Next steps (not necessarily in this order):
- add the stand-alone sideload and append funcs to the new type
- move this type to a package

**Review by tbg — APPROVED** [review]
> Reviewed 12 of 12 files at r5, 2 of 2 files at r6, 6 of 6 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91742)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pavelkalinnikov)

___
*[`pkg/kv/kvserver/replica_raft.go` line 1082 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGbK0YOE_O-Fu5m9a-f:-NGc7ccg5PA5SmzhoyHi:bz6q6fi) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bab8518c963da95cfe716b803aea07471e16289/pkg/kv/kvserver/replica_raft.go#L1082)):*
> I think it's a bit early and unclear now.

Sure, works for me as is. I always prefer not putting constructors in because then you're either passing an opaque function or some interface, and that's often more opaque/annoying. We can revisit later.
The raft log update, if it really uses blind writes, doesn't compose well with other operations anyway, so maybe having it use a batch that cannot be shared is the right move. We will see.
___
*[`pkg/kv/kvserver/replica_raft.go` line 1017 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGc6Wdb1S2FGHW3u1bi:-NGc6Wdb1S2FGHW3u1bj:bbsfbhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b169b353ce74235e4676d7fd012606636a4042f/pkg/kv/kvserver/replica_raft.go#L1017)):*
> ```Go
> 			return stats, expl, err
> 		}
> 		if state.raftLogSize -= purgedSize; state.raftLogSize < 0 {
> ```

I personally avoid these inline updates. Yes, it saves a line, but it also takes an update that happens unconditionally into a scope where most readers assume a temp assignment that doesn't affect outer scope.
Added a fixup commit in case you're amenable to that reasoning.
___
*[`pkg/kv/kvserver/replica_raft.go` line 940 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGc8ZzBAR4AGexMqYJO:-NGc8ZzBAR4AGexMqYJP:b-53xh9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3b1886db24874ce365fcf401838eb781509a16/pkg/kv/kvserver/replica_raft.go#L940)):*
> ```Go
> 	}
> 	// TODO(pavelkalinnikov): construct and store this in Replica.
> 	s := logStore{
> ```

Does `s` escape to the heap here?

You can find out with `./scripts/goescape ./pkg/kv/kvserver/replica_raft.go`.

If it does escape, we want to address the TODO before merging, to avoid an extra alloc in a hot path.
___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 683 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGbJ33N8BZVgHx9qjBB:-NGc7vn92dv4FB2sdBve:b-pcw3wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1c219f2e5c161252787e0c560aca8cb67ff90a3/pkg/kv/kvserver/replica_raftstorage.go#L683)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Good catch, fixed.
</blockquote></details>

Added a fixup commit with my suggestion above, which I still prefer, your call.

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91742)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/replica_raft.go` line 1017 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGc6Wdb1S2FGHW3u1bi:-NGcHAEWBKOYXxzZbmGB:b4mukws) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b169b353ce74235e4676d7fd012606636a4042f/pkg/kv/kvserver/replica_raft.go#L1017)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I personally avoid these inline updates. Yes, it saves a line, but it also takes an update that happens unconditionally into a scope where most readers assume a temp assignment that doesn't affect outer scope.
Added a fixup commit in case you're amenable to that reasoning.
</blockquote></details>

Good point, merged the fixup in.
___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 683 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGbJ33N8BZVgHx9qjBB:-NGcHLmA1-sXLanV6WO9:b6yfibf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1c219f2e5c161252787e0c560aca8cb67ff90a3/pkg/kv/kvserver/replica_raftstorage.go#L683)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Added a fixup commit with my suggestion above, which I still prefer, your call.
</blockquote></details>

SGTM, merged in.

**Review by pav-kv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91742)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/replica_raft.go` line 940 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/91742#-NGc8ZzBAR4AGexMqYJO:-NGcIcifBIFjKs6-ZV3T:b-8kzhd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f3b1886db24874ce365fcf401838eb781509a16/pkg/kv/kvserver/replica_raft.go#L940)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Does `s` escape to the heap here?

You can find out with `./scripts/goescape ./pkg/kv/kvserver/replica_raft.go`.

If it does escape, we want to address the TODO before merging, to avoid an extra alloc in a hot path.
</blockquote></details>

It doesn't. Nice tool.

**Comment by pavelkalinnikov** [comment]
bors r=tbg

**Comment by pavelkalinnikov** [comment]
bors cancel

**Comment by pavelkalinnikov** [comment]
Don't like the failing `TestJobsProtectedTimestamp` (not in all builds though). Holding off the merge.

**Comment by pavelkalinnikov** [comment]
`TestJobsProtectedTimestamp` is pre-existing. Merging.

bors r=tbg

**Pull request merged**
